### PR TITLE
HelpURL method in IAlgorithrm

### LINKED ISSUE #19521: HelpURL method in IAlgorithrm
Add virtual std::string HelpURL() method in IAlgorithm, default implementation is empy string. Later to be exposed to python. If overwritten, will be invoked by the help button in the algorithm dialog. Useful when the algorithm is not part of the Mantid distribution.

### PR DESCRIPTION
**Pull request by jmborr**
Description of work.

**To test:** Load this python algorithm in MantidPlot:
```
class TestHelpURL(PythonAlgorithm):
    def helpURL(self):
        return 'http://goo.gl/ccf4r3'
    def name(self):
        return 'TestHelpURL'
    def category(self):
        return 'misc'
    def PyInit(self):
        pass
    def PyExec(self):
        pass
AlgorithmFactory.subscribe(TestHelpURL)
```
Open the algorithm dialog and click on the help button, the help page will open in the default web browser.

Fixes #19521.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by gvardany** [comment]
Works fine if the url is correct. Otherwise it still tries to point to the regular documentation. Try this:
``` 
def helpURL(self):
    return 'google.com'  # not a full URL
```
OR
``` 
def helpURL(self):
    return 'something'  # not a URL
```

So I wonder what is the best place to validate it, and what is the best thing to do, if it is invalid.

File: MantidQt/MantidWidgets/src/MantidHelpWindow.cpp [hunk 1]
```diff
@@ -5,6 +5,7 @@
 #include "MantidKernel/ConfigService.h"
 #include "MantidKernel/Logger.h"
 #include "MantidKernel/RegistrationHelper.h"
+#include "MantidAPI/AlgorithmManager.h"
```
Comment (gvardany) [inline MantidQt/MantidWidgets/src/MantidHelpWindow.cpp]: Please sort these alphabetically

File: Framework/API/inc/MantidAPI/Algorithm.h [hunk 2]
```diff
@@ -200,6 +200,11 @@ class MANTID_API_DLL Algorithm : public IAlgorithm,
   /// is provided
   const std::string alias() const override { return ""; }
 
+  /// function to return URL for algorithm documentation; A default
+  /// implementation
+  /// is provided. Override if algorithm not part of the Mantid distribution.
```
Comment (gvardany) [inline Framework/API/inc/MantidAPI/Algorithm.h]: if algorithm is not a part of

File: Framework/API/inc/MantidAPI/IAlgorithm.h [hunk 3]
```diff
@@ -80,6 +80,10 @@ class MANTID_API_DLL IAlgorithm : virtual public Kernel::IPropertyManager {
   /// function to return any aliases of the algorithm.
   virtual const std::string alias() const = 0;
 
+  /// function to return optional URL for algorithm documentation.
+  /// Override if algorithm not part of the Mantid distribution
```
Comment (gvardany) [inline Framework/API/inc/MantidAPI/IAlgorithm.h]: if algorithm is not a part of

File: docs/source/release/v3.10.0/framework.rst [hunk 4]
```diff
@@ -17,6 +17,8 @@ Algorithms
 ----------
 
 - Removed the optional flag ``LocationParameters`` from ``ClearInstrumentParameters``.
+- Override method `helpURL` to return the webpage for the algorithm's documentation. Useful when registering Python
```
Comment (gvardany) [inline docs/source/release/v3.10.0/framework.rst]: `Algorithm::helpURL`

Comment (jmborr) [inline docs/source/release/v3.10.0/framework.rst]: it's not an Algorithm, it's just a method of class IAlgorithm.

File: Framework/API/inc/MantidAPI/IAlgorithm.h (hunk 3 above)
Comment (jmborr) [inline Framework/API/inc/MantidAPI/IAlgorithm.h]: ?

File: Framework/API/inc/MantidAPI/Algorithm.h (hunk 2 above)
Comment (jmborr) [inline Framework/API/inc/MantidAPI/Algorithm.h]: ?

File: docs/source/release/v3.10.0/framework.rst (hunk 4 above)
Comment (gvardany) [inline docs/source/release/v3.10.0/framework.rst]: Well that is clear, but from the sentence it is not obvious that it is a method of the base algorithm class, that could be overridden.

File: Framework/API/inc/MantidAPI/Algorithm.h (hunk 2 above)
Comment (gvardany) [inline Framework/API/inc/MantidAPI/Algorithm.h]: this was just a remark on english, seems the verb is missing in the second sentence

**Comment by jmborr** [comment]
@gvardany All URLs to be opened by the browser seem to be handled in the end by `MantidDesktopServices::openUrl(const QUrl &url)` which is a wrapper for `Qt::QDesktopServices::openUrl`. How about doing the validation here?

**Comment by peterfpeterson** [comment]
I think the only check you could do (which wouldn't be much) is to confirm that the url started with `http` (to allow for both `http` *and* `https`).

**Comment by jmborr** [comment]
@peterfpeterson I was lucky and QUrl::fromUserInput constructor fix this issue :smile:

**Comment by jmborr** [comment]
@gvardany Can I get the squirrel stamp? I want this to go to the 3.10 release.

**Comment by gvardany** [comment]
I think now it makes more sense, what it does. 
Just one issue, I get a segmentation fault when closing mantid, I am not sure though if it is related to the changes here, try this:
1. Open MantidPlot
2. Run the script to register the empty algorithm
3. Open it's auto generated dialog
4. Close MantidPlot
This does not happen if you have a dialog of a native mantid algorithm open.

**Comment by jmborr** [comment]
@gvardany  I did the following:
1. If you close the `TestHelpURL` dialog before closing Mantid, you will not get the segmentation fault.
2. If instead of `TestHelpURL` you open the dialog for any other Python algorithm, e.g `ABins`, and then close Mantid, you will get a segmentation fault.

In summary, this segfault affects all Python algorithms. You should open an issue.

**Comment by gvardany** [comment]
I have created the issue #19669, this one can be :shipit:

**Pull request merged**